app/browser/reducers/passwordManagerReducer.js

```diff
@@ -86,178 +87,145 @@ const init = () => {
-    let result = passwords.findLast((pw) => {
-      return pw.get('origin') === origin && pw.get('action') === action && (username ? pw.get('username') === username : !pw.get('username'))
+const migrate = (state) => {
+  masterKey = masterKey || getMasterKey()
+  if (!masterKey) {
+    console.log('Could not access master password; aborting')
+    return
+  }
+  const passwords = state.get('passwords')
+  if (passwords.size) {
+    passwords.forEach((password) => {
+      let decrypted = CryptoUtil.decryptVerify(password.get('encryptedPassword'),
+                                               password.get('authTag'),
+                                               masterKey,
+                                               password.get('iv'))
+      let form = {}
+      form['origin'] = password.get('origin')
+      form['signon_realm'] = password.get('origin') + '/'
+      form['action'] = password.get('action')
+      form['username'] = password.get('username')
+      form['password'] = decrypted
+      autofill.addLogin(form)
     })
-    if (result && password === CryptoUtil.decryptVerify(result.get('encryptedPassword'),
-                                                        result.get('authTag'),
-                                                        masterKey,
-                                                        result.get('iv'))) {
-      return
-    }
-
-    var message = username
-      ? locale.translation('notificationPasswordWithUserName').replace(/{{\s*username\s*}}/, username).replace(/{{\s*origin\s*}}/, origin)
-      : locale.translation('notificationPassword').replace(/{{\s*origin\s*}}/, origin)
-
-    if (!(message in passwordCallbacks)) {
-      // Notification not shown already
-      appActions.showNotification({
-        buttons: [
-          {text: locale.translation('yes')},
-          {text: locale.translation('no')},
-          {text: locale.translation('neverForThisSite')}
-        ],
-        options: {
-          persist: false,
-          advancedText: locale.translation('notificationPasswordSettings'),
-          advancedLink: 'about:passwords'
-        },
-        message
-      })
-    }
-
-    passwordCallbacks[message] = (buttonIndex) => {
-      delete passwordCallbacks[message]
-      appActions.hideNotification(message)
-
-      if (buttonIndex === 1) {
-        return
-      }
-      if (buttonIndex === 2) {
-        // Never save the password on this site
-        appActions.changeSiteSetting(origin, 'savePasswords', false)
-        return
-      }
-
-      // Save the password
-      const encrypted = CryptoUtil.encryptAuthenticate(password, masterKey)
-      appActions.savePassword({
-        origin,
-        action,
-        username: username || '',
-        encryptedPassword: encrypted.content,
-        authTag: encrypted.tag,
-        iv: encrypted.iv
-      })
-    }
-  })
+    state = state.set('passwords', new Immutable.List())
```

Comment:
it might be a good idea to copy the old passwords somewhere instead of deleting them forever. if the migration doesn't work for some reason, users will be mad if they can't get their passwords back.